File: khalti-android/src/main/java/khalti/checkOut/Wallet/WalletContract.java
package khalti.checkOut.Wallet;


import java.util.HashMap;

import khalti.checkOut.Wallet.helper.WalletConfirmPojo;
import khalti.checkOut.Wallet.helper.WalletInitPojo;
import khalti.checkOut.api.Config;
import rx.Observable;

public interface WalletContract {
    interface View {

        void toggleProgressDialog(String action, boolean show);

        void toggleConfirmationLayout(boolean show);

        void togglePinMessage(boolean show);

        void setPinMessage(String message);

        void setMobile(String mobile);

        HashMap<String, Observable<CharSequence>> setEditTextListener();

        void setEditTextError(String view, String error);

        void setButtonText(String text);

        Observable<Void> setButtonClickListener();

        Observable<Void> setImageClickListener();

        void setConfirmationLayoutHeight(String view);

        void showNetworkError();

        void showMessageDialog(String title, String message);

        void showPINDialog(String title, String message);

        void showPINInBrowserDialog(String title, String message);

        void openKhaltiSettings();

        void openLinkInBrowser(String link);

        void closeWidget();

        void updateConfirmationHeight();

        String getMessage(String action);

        boolean hasNetwork();

        String getPayButtonText();

        HashMap<String, String> getFormData();

        void showSlogan();

        void showBranding();

        void setPresenter(Presenter presenter);
    }

    interface Presenter {

        void onCreate();

        void onDestroy();

        void openKhaltiSettings();

        void openLinkInBrowser();

        void showPINInBrowserDialog(String title, String message);

        boolean isMobileValid(String mobile);

        boolean isFinalFormValid(String pin, String confirmationCode);

        void initiatePayment(boolean isNetwork, String mobile);

        void confirmPayment(boolean isNetwork, String confirmationCode, String transactionPin);

        void updateConfirmationHeight();

        void unSubscribe();
    }

    interface Model {

        Observable<WalletInitPojo> initiatePayment(String mobile, Config config);

        Observable<WalletConfirmPojo> confirmPayment(String confirmationCode, String transactionPIN);

        void unSubscribe();
    }
}

File: khalti-android/src/main/java/com/khalti/checkOut/wallet/WalletContract.kt
package com.khalti.checkOut.wallet


import com.khalti.base.LifeCycle

import com.khalti.checkOut.wallet.helper.WalletConfirmPojo
import com.khalti.checkOut.wallet.helper.WalletInitPojo
import com.khalti.checkOut.api.Result
import com.khalti.checkOut.helper.Config
import com.khalti.signal.Signal

interface WalletContract {
    interface View {

        val payButtonText: String

        val formData: Map<String, String>

        fun toggleProgressDialog(action: String, show: Boolean): Signal<Boolean>?

        fun toggleConfirmationLayout(show: Boolean)

        fun togglePinMessage(show: Boolean)

        fun setPinMessage(message: String)

        fun setMobile(mobile: String)

        fun setEditTextError(view: String, error: String?)

        fun setButtonText(text: String)

        fun setConfirmationLayoutHeight(view: String)

        fun setEditTextListener(): Map<String, Signal<CharSequence>>

        fun setClickListener(): Map<String, Signal<Any>>

        fun showNetworkError()

        fun showMessageDialog(title: String, message: String)

        fun showPINDialog(title: String, message: String): Signal<Boolean>

        fun showSlogan()

        fun showBranding()

        fun openKhaltiSettings()

        fun openLinkInBrowser(link: String)

        fun closeWidget()

        fun updateConfirmationHeight()

        fun getMessage(action: String): String

        fun hasNetwork(): Boolean

        fun doesPackageExist(): Boolean

        fun setPresenter(presenter: Presenter)
    }

    interface Presenter : LifeCycle {

        fun isMobileValid(mobile: String): Boolean

        fun isInitialFormValid(mobile: String, pin: String): Boolean

        fun isFinalFormValid(pin: String, confirmationCode: String): Boolean

        fun onInitiatePayment(isNetwork: Boolean, mobile: String, pin: String)

        fun onConfirmPayment(isNetwork: Boolean, confirmationCode: String, transactionPin: String)
    }

    interface Model {

        suspend fun initiatePayment(mobile: String, pin: String, config: Config): Result<WalletInitPojo>

        suspend fun confirmPayment(confirmationCode: String, pin: String, token: String): Result<WalletConfirmPojo>
    }
}
